### ISSUE TITLE
Support different submission scenarios (start/continue/restart)

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-web#440]
The V2 workflows seem to be very optimized for continuing workflows from where one has stopped. But another typical behavior is that a user reruns the whole workflow from the beginning.

That gets confusing, because there can already be an existing image list and existing filters that are set. For example, if a user runs a workflow that goes Convert OME-Zarr, MIP.
Then they want to rerun it, it fails. Because the is_3D filter was set to False and rerunning Convert OME-Zarr does not change the filter.

I haven't fully figured out what we'll need to handle this better. Here are a few points:

- We'll need to expose what filters are already active when a workflow gets submitted.
- I think we may need different concepts for rerunning a workflow from the beginning vs. continuing an existing workflow

Rerunning a workflow: 
- What happens with the existing image list?
- What happens with the existing filters?

**Comment by jluethi** [comment]
2 issues to address some of it by making filters more visible & editable:
https://github.com/fractal-analytics-platform/fractal-web/issues/441
https://github.com/fractal-analytics-platform/fractal-web/issues/442

Let's start with that and brainstorm how we handle full reruns of workflows based on that

**Comment by jluethi** [comment]
_brainstorming notes_
Another part to consider here:
Does this depend on whether the workflow is run from the first task or an intermediary task?

Maybe we have different submissions in the web interface for that:
- Initial run: Select dataset and until which task you want to run
- Rerun workflow: Resets filters & image list, always starts from the beginning until the task that is set
- Continue workflow (or better word): Start from a user selected task with the current filters [show them] until a given task. Could be from the task after the last successful task, or from an arbitrary task

I'm mostly thinking about this as a way the web portal offers different submission strategies (and calls one or multiple APIs depending on the selection), not as new backend features.

**Comment by jluethi** [comment]
I thought about this some more and tested additional workflows. It's something we need to address. Here's a proposal on how we address it now (part 1) & where we could go in the future (part 2).

----

# Part 1: What we should do now

We should split the submission into different categories, as outlined in the comment above:

### i) No workflow has been run:
We do basically what we've done so far. But let's set the first task of the workflow as default "First task". (depending on what we do in iii, we could also make the "First task" something the user can't change anymore, it's just shown that the workflow will start there, as it's the first task in the workflow).

<img width="1329" alt="Screenshot 2024-04-17 at 18 36 39" src="https://github.com/fractal-analytics-platform/fractal-web/assets/18033446/1e66186e-3700-43fd-b5a1-2e2bf7ae4ff0">


### ii) Workflow has been run & I want to continue
This is how the submission currently behaves when a workflow has already been run. Let's rename the button to Continue Workflow if a workflow has already been run.

<img width="809" alt="Screenshot 2024-04-17 at 18 37 54" src="https://github.com/fractal-analytics-platform/fractal-web/assets/18033446/0f616625-9300-4118-bc76-cba8341a2d58">

This just submits the workflow as we're doing so far.

<img width="321" alt="Screenshot 2024-04-17 at 18 41 59" src="https://github.com/fractal-analytics-platform/fractal-web/assets/18033446/8004bc3c-6942-49ae-a569-666bb3e6dc9e">

Best-case, we would make the First Task default to 
a) The task that failed if any task failed
b) the next task after the last one that finished successfully
c) The last task if there are no unfinished tasks (?)

The user can then change which task is chosen there, e.g. continue the workflow by rerunning a task that has already ran before.

This likely depends on https://github.com/fractal-analytics-platform/fractal-server/issues/1083 and may not be trivial. In that case, the defaults for "First task" can also be moved to a future goal (part 2)

### ii) Workflow has been run & I want to start over from the beginning

This is currently not supported. One way we can support this:
Add a "Rerun workflow from the beginning" button. It has the following behavior:
1) It deletes the image list
2) It removes all filters
3) It submits the workflow from the first task

<img width="320" alt="Screenshot 2024-04-17 at 18 42 09" src="https://github.com/fractal-analytics-platform/fractal-web/assets/18033446/660caac8-4d44-4001-93dc-196810c5acbd">

It should come with a big warning saying something like:

> ! Warning !  Rerunning a workflow from the beginning will remove all items from the image list and reset all filters. It will not remove any of the actual image data though.

Also, to avoid users using this option for the wrong thing (e.g. my task 7 failed, let me rerun the workflow), I'd say the "First task" should be selected as the first task of the workflow and be unchangeable (=> it just shows that it will start in the beginning with no option for the user to change that).


I'm open to different suggestions for naming the button, but quite convinced that we should have these 3 distinct modes.


---------

# Part 2: Goals we may achieve later

Potentially defaults for Continue workflow will go into this category

Also, continue workflow would benefit a lot from showing something like:
This workflow will continue with the following image list, these filters are set.

But we can add that feature later.

**Comment by jluethi** [comment]
I think this is mostly something we can solve on the front-end, with 2 areas that would involve new server functionality:
1. Finding the right "default start task" for the Continue option
2. Easy way to delete the whole image list & remove all filters?

**Comment by tcompa** [comment]
Minor comment:

> -     It deletes the image list
> -     It removes all filters
> -     It submits the workflow from the first task

This list should also include the dataset history, right?

If not, it will appear as if some of the workflow tasks already ran, and they will have a done/failed status.

**Comment by tcompa** [comment]
Another remark - cc @jluethi

In the scenario 1.iii (that is, where I want to restart from the beginning with an empty dataset), is there any value in emptying an existing dataset rather than just creating a new one?

If this is made to simplify the user flow (that is, to avoid that they need to go back to the dataset page), this feature can be embedded within the restart-workflow-from-the-beginning modal. The zarr-dir could be prefilled from the original one, and the dataset-name could be the same as before with a suffix. TBD

**Comment by jluethi** [comment]
tl;dr: I think the dataset approach should work. Let me know what you think.
I'll sketch it in more detail tomorrow morning :)

**Comment by jluethi** [comment]
I think the dataset approach can work like this:
If a user clicks "Rerun workflow from the beginning", they see a warning message that says:

> ! Warning !  Rerunning a workflow from the beginning will create a new dataset and optionally delete the old dataset with its image list, filters & history.

By default, a checkbox is clicked to have the old dataset replaced (e.g. delete the old dataset, create a new one with the same name as the deleted one & give that new dataset the same zarr_dir).

<img width="408" alt="Screenshot 2024-04-19 at 08 44 56" src="https://github.com/fractal-analytics-platform/fractal-web/assets/18033446/a75b6894-f694-4940-a84e-4f5cb45a33f3">

If the user unticks the `Replace existing dataset` box, a new entry shows below the dataset selector for "New dataset name", which defaults to the existing dataset name + suffix (e.g. _1, _2, _n+1).
If `Replace existing dataset` is unchecked, then the old dataset does not get deleted, but a new dataset with the same zarr_dir is still created

**Comment by tcompa** [comment]
Here is our plan with @zonia3000.

# Preliminary changes

These changes would make the following easier:

1. When no dataset is selected, and the "submit" button is clicked, show an error message like "you should first select a dataset".
2. Given 1, then we can make the dataset selection in the submission modal non-editable. Note: we still always display it (for any of the submission scenarios).

cc @jluethi are these two minor changes OK?

## Has any job run?

Given the previous section, we know workflow_id and dataset_id.

We then call
```
GET /api/v2/project/{project_id}/status/?dataset_it={dataset_id}&workflow_id={workflow_id}
```
and if the response is `{"status": {}}` then we set `HasAnyJobRun=false`, otherwise it is true.

Note: if we hit any edge case here, we could switch to using `GET /api/v2/project/{id}/workflow/{id}/job/`

## If `HasAnyJobRun=false`

We show a "submit" button, which opens a modal similar to the current one.

Differences:

1. The dataset is non-editable (see above)
2. The first task is set to match with the first task in the workflow (editable, for now)

## If `HasAnyJobRun=true`

We show two buttons: "continue" or "restart"

### If "continue" is clicked

We show a submission modal similar to the current one, possible with a pre-selected first task.
Let's postpone the definition of this first task to a different issue (ref XXX).

### If "restart" is clicked

First, we show a big warning with text

> WARNING: Rerunning a workflow from the beginning will create a new dataset and optionally delete the old dataset with its image list, filters & history.

Then the user can uncheck a "replace existing dataset" (checked by default)

#### If "replace dataset" is checked

Show original dataset name (non editable).
Store current dataset `name` and `zarr_dir`.

Replace dataset with new one. For the moment this should go through
DELETE original dataset
POST new dataset with same `name` and `zarr_dir` (and no other property).
(but maybe we'll introduce a specific PATCH in the future)

Update the currently-selected dataset ID in workflow-page and submission modal (same variable).

#### If "replace dataset" is unchecked

Show original dataset name (non editable).
Show "new dataset name" editable prompt, with a default (see below)
POST new dataset with new `name` and old `zarr_dir` (and no other property).

How to set the default for new dataset name?
Prefix is the current name.
Existing names are obtained via `GET /project/{id}/dataset/`.
We iterate over {prefix}_1, {prefix}_2, .. until we find one that doesn't exist yet.

**Comment by jluethi** [comment]
Looks great overall!

A few minor comments:

> 1. When no dataset is selected, and the "submit" button is clicked, show an error message like "you should first select a dataset".
> 2. Given 1, then we can make the dataset selection in the submission modal non-editable. Note: we still always display it (for any of the submission scenarios).

Why don't we make the run button greyed out / unclickable then when no dataset is selected? With the work in https://github.com/fractal-analytics-platform/fractal-web/issues/445, no dataset selection will eventually anyway become rarer and we'd avoid having to have an error message dialogue

+1 on the making the dataset something the user can't change, but is still shown.

----

Flow for `HasAnyJobRun=false`: Sounds good! On the naming of the button: "Run workflow" seems clearer than "submit" to me.

Flow for `continue`: Sounds good, fine with postponing the first task selection choice. Let's call it "Continue Workflow"

On `restart`:
Let's call it "Restart workflow"

If "replace dataset" is checked => fine with that flow!

If "replace dataset" is unchecked:
Any good reason for using prefixes instead of suffixes? I like the logic for choosing them. Would have intuitively gone with suffixes, as a user could have e.g.
another_test_dataset
my_core_dataset
some_subset

In that case, adding prefixes groups them by prefix when sorted alphabetically, which isn't very meaningful. Suffixes would have them sorted by the core name

**Comment by tcompa** [comment]
> Why don't we make the run button greyed out / unclickable then when no dataset is selected?

That is also fine, but if we only make the button unclickable then we are not telling the user how to proceed. In that case do we need to add a warning message that is always displayed when no dataset is selected?

---

As for the renaming of different components -> sure, thanks. Our names were meant to be placeholders.

---

> Any good reason for using prefixes instead of suffixes? I like the logic for choosing them. Would have intuitively gone with suffixes, as a user could have e.g.

Then I may have misuderstood the previous comment:

> "New dataset name", which defaults to the existing dataset name + suffix (e.g. _1, _2, _n+1).

Let's make this more explicit:
If there exists a single dataset and it named "MyDataset", what is the default for the new name? My guess is that we want to set it to "MyDataset_1", but if you prefer "1_MyDataset" there is no problem.

**Comment by jluethi** [comment]
> That is also fine, but if we only make the button unclickable then we are not telling the user how to proceed. In that case do we need to add a warning message that is always displayed when no dataset is selected?

Given that in the near future, there will by default always be a dataset selected, I think that's not necessary. And the selector is right next to the run button.

> If there exists a single dataset and it is named "MyDataset", what is the default for the new name?

My bad, I misunderstood. Yes, the dataset name is the prefix, the number is the suffix.

> My guess is that we want to set it to "MyDataset_1", but if you prefer "1_MyDataset" there is no problem.

Yes, I'd want "MyDataset_1". Didn't read your message above carefully enough

**Comment by jluethi** [comment]
Good idea. The warning message could be the one drafted above extended:

> WARNING: Restarting a workflow will create a new dataset and optionally delete the old dataset with its image list, filters & history. It does not remove the existing Zarr data though. Either remove it yourself before rerunning or set the overwrite option to True in the corresponding tasks.


(The way users usually get around this when rerunning workflows in V1 is setting overwrite to True and that works robustly)

**Comment by jluethi** [comment]
And the issue that rerunning a workflow without overwrite = True fails is something we already have and where we provide informative error messages that guide users to setting overwrite=True in many of our tasks

**Comment by zonia3000** [comment]
Run workflow button disabled if no dataset is selected:

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/49a4b43f-8a10-4ff0-9857-ace63977675d)

If a previous job was run the 2 buttons "Continue workflow" and "Restart workflow" are shown:

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/ec5fc19a-06df-4e2c-966c-14dfac9393c5)

Clicking on "Continue workflow":

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/9a57cd5f-12f2-4134-8c7b-d6e89352a1eb)

Restart workflow with "Replace existing dataset" checked:

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/933bba93-1b50-43e0-a525-a29a5a23a596)

Restart workflow with "Replace existing dataset" unchecked:

![image](https://github.com/fractal-analytics-platform/fractal-web/assets/5459334/b0d4ab3e-e026-4372-acd2-df713abbb252)

**Comment by tcompa** [comment]
Nice!

Very quick comment: maybe the "restart workflow modal should include a pre-selected first task, set to the first task in the workflow?
